src/refine_base/refine_base.1
.TH refine_base 1 "April 2004"

.SH NAME:
refine_base \- perform baseline refinement.

.SH SYNOPSIS:
.B "refine_base" 
.BI "[-keep <iter>] [-log <file>] [-quiet]"
.br
.BI "          <phase> <tie_points> <meta> <old_base> <new_base>"

.SH DESCRIPTION:
Refine_base corrects a given baseline for an interferogram using tie points.
Tie points are listed in an ASCII file with each tie point on a separate line.
The format for a line in a tie point file is x or sample, y or line, and height
in meters. These coordinates are before multilooking, and in the original
interferogram's image space (before any windowing or multilooking). X and Y are
pixel locations in the file and each  parameter is separated by white space.
.PP
Using the tie point file, refine_base retrieves the unwrapped phase value at
each pixel. If the phase value happens to equal the seed value used to unwrap
the interferogram, that point is discarded. Refine_base will assume a default
seed value of 0.0. You should change this to the value used in our progam
escher(1).
.PP
After retrieving the phase values, refine_base calculates the new baseline
values and places the resulting information in the new_base file.
.PP
The baseline file contains 4 values describing the baseline separation of the
satellite passes. The 4 values are, in order, perpendicular baseline (Bn), 
delta perpendicular baseline (dBn), parallel baseline (Bp), and delta parallel 
baseline (dBp). The delta values represent the change of the baseline over the
entire image. All these values are in meters.

.SH COMMAND LINE OPTIONS:
.TP 8
.B "-keep <iter>"
Keep intermediate files generated by refine_base. By default, refine_base will
delete all the intermediate files it creates. However, if you wish to save
these files, specify the \-keep option and a value for <iter>. The intermediate
files will be appended with this value.
.TP 8
.B "-log <file>"
Writes output to a log <file> (and the terminal).
.TP 8
.B "-quiet"
Surpress output to only the essential.
.TP 8
.B "phase"
Input unwrapped phase file. This file is usually generated by the escher(1)
program.
.TP 8
.B "tie_points"
An ASCII tie point file containg x,y image locations and height values in
meters. 
.TP 8
.B "meta"
Name of .meta file for the source image.
.TP 8
.B old_base
Baseline file. ASCII file of 4 baseline parameters entered on the same line.
This file contains the baseline parameters that need to be refined.
.TP 8
.B new_base
Baseline file generated by refine_base. This file contains the new baseline
parameters. The interferogram should be deramped again using these values.

.SH EXAMPLES:
.ce 1
example% refine_base uwp12 ifm12.tp image1 base.0 base.1
.PP
Refine_base will refine the baseline in the file base.0 and place the new
baseline in the file base.1. It will use the tie points in the file ifm12.tp
and obtain the phases from the file uwp12.phase. Image1 is the base name of the
metadata file image1.meta associated with the source image. 

.SH SEE ALSO:
coregister_coarse(1), deramp(1), escher(1)

.SH HARDWARE/SOFTWARE LIMITATIONS:
None known.

.SH BUGS:
None known.

.SH AUTHOR:
Mike Shindle,   6/97
.br
Patrick Denny,  4/04

.SH CONTACT:
uso@asf.alaska.edu
.br
http://www.asf.alaska.edu




src_ifm/refine_base/refine_base.c
/*************************************************************************
 NAME:  refine_base

 SYNOPSIS:  refine_base [-k __iter__ ] [-log <file>] [-quiet] 
                        <phase> <tie_points> <meta> <old_base> <new_base>

  -k __iter__    Keep intermediate products.  Iterations=__iter__
  -log <file>	 Allows the output to be written to a log file
  -quiet	 Suppresses the output to the essential

       <phase>   unwrapped interferogram phase file" ;
  <tie_points>   tie-point location file.";
        <meta>   filename containing interferogram's metadata.";
    <old_base>   baseline file containg four parameters:" ;
                 Bn delta_Bn Bp delta_Bp" ;
    <new_base>   refined baseline file containg four parameters:" ;
                 Bn delta_Bn Bp delta_Bp" ;

 DESCRIPTION:
 
     Perform baseline refinement using a set of 4 functions: getphase, genab,
     bp, and test_base. The last one, test_base, does not refine the
     baseline at all. Its only purpose is to check the validity of the new
     baseline.

     Getphase extracts phase values from an associated tie point file. 

     Genab creates a matrix (A) and a vector (b) that are related by the
     equation Ax=b. 

     Bp solves the above equation in which the vector x is our new baseline
     values.


 EXTERNAL:

	bp


 PROGRAM HISTORY:
    VERS:   DATE:        PURPOSE:
    ---------------------------------------------------------------
    1.0	    11/96	 M. Shindle - Original program
    1.1	    8/97	 O. Lawlor  - Takes Parameters from metadata & ddr
    1.2	    7/01	 R. Gens    - Conversion of script into program
				      Included external programs as functions
    1.5    12/03         P. Denny   - Bring commandline parsing to our standard
                                      Use meta 1.1 instead of DDR (genab.c)

HARDWARE/SOFTWARE LIMITATIONS:

ALGORITHM DESCRIPTION:

ALGORITHM REFERENCES:

BUGS:
	
****************************************************************************/

/******************************************************************************
*                                                                             *
* Copyright (c) 2004, Geophysical Institute, University of Alaska Fairbanks   *
* All rights reserved.                                                        *
*                                                                             *
* You should have received an ASF SOFTWARE License Agreement with this source *
* code. Please consult this agreement for license grant information.          *
*                                                                             *
*                                                                             *
*       For more information contact us at:                                   *
*                                                                             *
*	Alaska Satellite Facility	    	                              *
*	Geophysical Institute			www.asf.alaska.edu            *
*       University of Alaska Fairbanks		uso@asf.alaska.edu	      *
*	P.O. Box 757320							      *
*	Fairbanks, AK 99775-7320					      *
*									      *
******************************************************************************/


#include "asf.h"

#define VERSION 1.5

/* Prototypes */
int getphase(char *phasein, char *tiept, char *outfile);
int genab(char *datafile, char *basefile, char *metaName, char *matfile,
          char *vecfile);
int bp(char *matfile, char *vecfile, char *newbase);
int test_base(char *basefile, char *matfile, char *vecfile);

static
void usage (char *name) {
 printf("\n"
 	"USAGE:\n"
	"   %s [-keep <iter>] [-log <file>] [-quiet]\n"
	"            <phase> <tie_points> <meta> <old_base> <new_base>\n",name);
 printf("\n"
	"REQUIRED ARGUMENTS:\n"
	"   phase        Unwrapped interferogram phase file\n"
	"   tie_points   Tie-point location file.\n"
	"   meta         Filename containing interferogram's metadata.\n"
	"   old_base     Baseline file containg four parameters:\n"
	"                  Bn delta_Bn Bp delta_Bp\n"
	"   new_base     Refined baseline file containg four parameters:\n"
	"                  Bn delta_Bn Bp delta_Bp\n");
 printf("\n"
	"OPTIONAL ARGUMENTS:\n"
	"   -keep <iter>  Keep intermediate products.  Iterations=<iter>\n"
	"   -log <file>   Allows the output to be written to a log file\n"
	"   -quiet	  Suppresses the output to the essential\n");
 printf("\n"
	"DESCRIPTION:\n"
	"   Corrects a given baseline for an interferogram using tie points.\n");
 printf("\n"
	"Version: %.2f, ASF SAR Tools\n"
	"\n", VERSION);
 exit(EXIT_FAILURE);
}

int main(int argc,char *argv[])
{
  int iter=0, keepIter=0;
  char cmd[255], phase_file[255], tp_file[255], ctrlpt_file[255];
  char meta_file[255], oldbase_file[255], newbase_file[255], matrix_file[255], vec_file[255];

  /* Parse command line arguments */
  logflag=quietflag=FALSE;
  while (currArg < (argc-5)) {
     char *key = argv[currArg++];
     if (strmatch(key,"-keep")) {
        CHECK_ARG(1);
	iter = atoi(GET_ARG(1));
	keepIter=TRUE;
     }
     else if (strmatch(key,"-log")) {
        CHECK_ARG(1);
        strcpy(logFile,GET_ARG(1));
        fLog = FOPEN(logFile, "a");
        logflag=TRUE;
     }
     else if (strmatch(key,"-quiet")) {
        quietflag=TRUE;
     }
     else {printf( "\n**Invalid option:  %s\n",argv[currArg-1]); usage(argv[0]);}
  }
  if ((argc-currArg) < 5) {printf("Insufficient arguments.\n"); usage(argv[0]);}

  /* take care of required command line arguments */
  strcpy(phase_file,   argv[currArg]);
  strcpy(tp_file,      argv[currArg+1]);
  strcpy(meta_file,    argv[currArg+2]); /* Do we need this argument?? */
  strcpy(oldbase_file, argv[currArg+3]);
  strcpy(newbase_file, argv[currArg+4]);

  printf("%s\n",date_time_stamp());
  printf("Program: refine_base\n\n");
  if (logflag) {
    StartWatchLog(fLog);
    printLog("Program: refine_base\n\n");
  }

  sprintf(ctrlpt_file, "ctrlpts.%d", iter);
  sprintf(matrix_file, "matrix.%d", iter);
  sprintf(vec_file, "vecB.%d", iter);

  /* get phases from unwrapped phase file */
  getphase(phase_file, tp_file, ctrlpt_file);

  /* generate matricies from phases & tie points*/
  genab(ctrlpt_file, oldbase_file, meta_file, matrix_file, vec_file);
 
  /* create baseline and test it */
  bp(matrix_file, vec_file, newbase_file);
  test_base(newbase_file, matrix_file, vec_file);

  /* remove intermediate files unless requested not to */
  if (keepIter==0) {
    sprintf(cmd, "rm %s %s %s", ctrlpt_file, matrix_file, vec_file);
    system(cmd);
  }
  
  return(0);
}

